exec/testing/fake_exec.go

```diff
@@ -45,6 +71,82 @@ func (fake *FakeExec) LookPath(file string) (string, error) {
 	return fake.LookPathFunc(file)
 }
 
+func getCaller(n int) string {
+	_, file, line, ok := runtime.Caller(n + 1)
+	if !ok {
+		return "???"
+	}
+	if i := strings.LastIndexAny(file, "/\\"); i >= 0 {
+		file = file[i+1:]
+	}
+	return fmt.Sprintf("%s:%d", file, line)
+}
+
+func (fake *FakeExec) fail(format string, args ...interface{}) {
+	msg := fmt.Sprintf(format, args...)
+	if fake.T == nil {
+		panic(msg)
+	}
+
+	// Don't re-fail from a "defer fexec.AssertExpectedCommands()" if we already failed once
+	if fake.T.Failed() {
+		return
+	}
+	fake.T.Fatal(msg)
+}
+
+// AddCommand adds a Cmd to be returned by a future fake.Command(...) call. Call
+// SetCombinedOutput or SetRunOutput on the result to specify its result.
+func (fake *FakeExec) AddCommand(cmd string, args ...string) *FakeCmd {
+	expectCaller := getCaller(1)
+	expectedArgv := append([]string{cmd}, args...)
+	fcmd := &FakeCmd{}
+	fake.CommandScript = append(fake.CommandScript,
+		func(cmd string, args ...string) exec.Cmd {
+			InitFakeCmd(fcmd, cmd, args...)
+			if !reflect.DeepEqual(expectedArgv, fcmd.Argv) {
+				fake.fail("Wrong command: expected\n%v (at %s)\ngot\n%v (at %s)", expectedArgv, expectCaller, fcmd.Argv, getCaller(2))
+			}
+			return fcmd
+		},
+	)
+	return fcmd
+}
+
+func outputAsBytes(output interface{}) []byte {
+	if outputBytes, ok := output.([]byte); ok {
+		return outputBytes
+	} else if outputStr, ok := output.(string); ok {
+		return []byte(outputStr)
+	} else if output == nil {
+		return nil
+	}
+	panic("output must be []byte or string")
+}
+
+// SetRunOutput provides the result of calling Run() on a FakeCmd.
+// stdout and stderr can be either []byte or string (which will be converted to []byte).
+func (fcmd *FakeCmd) SetRunOutput(stdout, stderr interface{}, err error) {
+	fcmd.RunScript = []FakeRunAction{
+		func() ([]byte, []byte, error) { return outputAsBytes(stdout), outputAsBytes(stderr), err },
+	}
+}
+
+// SetCombinedOutput provides the result of calling CombinedOutput() on a FakeCmd.
+// output can be either a []byte or a string (which will be converted to a []byte).
+func (fcmd *FakeCmd) SetCombinedOutput(output interface{}, err error) {
```

Comment:
Literally every existing caller in kubernetes wants to pass a string rather than a []byte, so if you don't want it to accept both then I'd change the API to just take a string instead, and say people can use the lower-level API if they need to use []byte